orderbook/src/database/app_data.rs

```diff
@@ -0,0 +1,225 @@
+use super::*;
+use anyhow::Result;
+use futures::stream::TryStreamExt;
+use futures::StreamExt;
+use model::app_data::AppDataBlob;
+use primitive_types::{H160, H256};
+use std::borrow::Cow;
+use thiserror::Error;
+
+#[async_trait::async_trait]
+pub trait AppDataStoring: Send + Sync {
+    async fn insert_app_data(&self, app_data: &AppDataBlob) -> Result<H256, InsertionError>;
+    fn app_data<'a>(&'a self, filter: &'a AppDataFilter) -> BoxStream<'a, Result<AppDataBlob>>;
+}
+
+/// Any default value means that this field is unfiltered.
+#[derive(Clone, Default)]
+pub struct AppDataFilter {
+    pub app_data_hash: Option<H256>,
+    pub referrer: Option<H160>,
+}
+
+#[derive(Error, Debug)]
+pub enum InsertionError {
+    #[error("Parsing the string was not successful:")]
+    ParsingStringError(#[from] serde_json::Error),
+    #[error("Duplicated record for `{0}`:")]
+    DuplicatedRecord(H256),
+    #[error("Anyhow error:")]
+    AnyhowError(#[from] anyhow::Error),
+    #[error("Database error:")]
+    DbError(#[from] sqlx::Error),
+}
+
+#[async_trait::async_trait]
+impl AppDataStoring for Postgres {
+    async fn insert_app_data(&self, app_data_str: &AppDataBlob) -> Result<H256, InsertionError> {
+        let app_data = app_data_str.get_app_data()?;
+        const QUERY: &str = "\
+            INSERT INTO app_data (\
+                app_data_hash, app_code, referrer, file_blob)\
+                VALUES ($1, $2, $3, $4);";
+        let app_data_hash = app_data_str.sha_hash()?;
+        sqlx::query(QUERY)
+            .bind(app_data_hash.clone().as_bytes())
+            .bind(
+                app_data
+                    .app_code
+                    .unwrap_or_else(|| "".to_string())
+                    .as_bytes(),
+            )
+            .bind(
+                app_data
+                    .metadata
+                    .unwrap_or_default()
+                    .referrer
+                    .unwrap_or_default()
+                    .address
+                    .as_ref(),
+            )
+            .bind(app_data_str.0.as_bytes())
+            .execute(&self.pool)
+            .await
+            .map(|_| ())
+            .map_err(|err| {
+                if let sqlx::Error::Database(db_err) = &err {
+                    if let Some(Cow::Borrowed("23505")) = db_err.code() {
+                        return InsertionError::DuplicatedRecord(app_data_hash);
+                    }
+                }
+                InsertionError::DbError(err)
+            })?;
+        Ok(app_data_hash)
+    }
+
+    fn app_data<'a>(&'a self, filter: &'a AppDataFilter) -> BoxStream<'a, Result<AppDataBlob>> {
+        if let Some(app_data_hash) = &filter.app_data_hash {
```

Comment:
This if/else statement is maybe not the most elegant way to implement these filters - especially if we add more and start mixing and matching. Maybe check how we do it for nullable filter fields in the [order](https://github.com/gnosis/oba-services/blob/42483040f4d51f2b74f309e4b4751955345c0c9b/orderbook/src/database/orders.rs#L172-L177) table?

Also right now you would implicitly ignore the referrer field if the app_data_hash filter was set (in which case an enum would be the better filter type).